### PR TITLE
[4854] Add RTE config to page builder

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/4854

### CODE REVIEW

File: ui/guest/src/store/epics/root.ts [hunk 1]
```diff
@@ -311,6 +311,8 @@ const epic: Epic<GuestStandardAction, GuestStandardAction, GuestState> = combine
         const draggable = ElementRegistry.getDraggable(record.id);
         const validations = field?.validations;
         const type = field?.type;
+        // @ts-ignore
+        const rteConfig = action.payload.rteConfig;
```
Comment (rart) [inline ui/guest/src/store/epics/root.ts]: What's the ts-ignore needed for?

File: ui/guest/src/react/Guest.tsx [hunk 2]
```diff
@@ -134,7 +135,7 @@ function Guest(props: GuestProps) {
               event.preventDefault();
               event.stopPropagation();
             }
-            dispatch({ type: type, payload: { event, record } });
+            dispatch({ type: type, payload: { event, record, rteConfig } });
```
Comment (rart) [inline ui/guest/src/react/Guest.tsx]: No need to send config here as the RTE config is already in the state.
No need for the const above either once this is removed.

File: ui/guest/src/store/epics/root.ts (hunk 1 above)
Comment (jvega190) [reply ui/guest/src/store/epics/root.ts]: not needed anymore, removing

File: ui/guest/src/controls/rte.ts [hunk 3]
```diff
@@ -152,7 +155,10 @@ export function initTinyMCE(
           e.preventDefault();
         }
       });
-    }
-  });
+    },
+    ...rteConfig
```
Comment (rart) [inline ui/guest/src/controls/rte.ts]: We must make sure that certain props prevail — like setup and possibly other important behavioural props — hence they must be after the `...rteConfig` statement

File: ui/guest/src/controls/rte.ts [hunk 4]
```diff
@@ -37,19 +39,23 @@ export function initTinyMCE(
...
     toolbar: type === 'html',
     menubar: false,
-    inline: true,
     base_url: '/studio/static-assets/modules/editors/tinymce/v5/tinymce',
     suffix: '.min',
+    ...rteConfig,
+    // Target can't be changed
+    target: record.element,
+    // Inline view doesn't behave well on pageBuilder, this setting shouldn't be changed.
+    inline: false,
```
Comment (rart) [inline ui/guest/src/controls/rte.ts]: Don't think this makes sense? This is backwards.

File: ui/guest/src/controls/rte.ts [hunk 5]
```diff
@@ -152,7 +158,19 @@ export function initTinyMCE(
...
-    }
-  });
+    },
+    // Autosave options are left as undefined since it is not supported in control.
+    autosave_ask_before_unload: undefined,
+    autosave_interval: undefined,
+    autosave_prefix: undefined,
+    autosave_restore_when_empty: undefined,
+    autosave_retention: undefined,
+    // No file picker is set by default, and functions are not supported in state.
```
Comment (rart) [inline ui/guest/src/controls/rte.ts]:
```suggestion
    // No file picker is set by default, and functions are not supported in config file.
```

File: ui/guest/src/controls/rte.ts [hunk 6]
```diff
@@ -152,7 +158,19 @@ export function initTinyMCE(
...
+    file_picker_callback: undefined,
+    // Height is set to the size of content
+    height: undefined
+  };
+
+  window.tinymce.init(settings);
```
Comment (rart) [inline ui/guest/src/controls/rte.ts]: What is settings var useful for? Why not leave with the single line init i.e. `init({ ourNonOverridableSettings, ...rteConfig, ourNonOverridableSettings })`?

**Review by rart — CHANGES_REQUESTED** [review]

**Comment by jvega190** [comment]
@rart changing to draft PR to wait for https://github.com/craftercms/studio-ui/pull/2458 to be merged so this can be updated

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 7]
```diff
@@ -455,7 +454,7 @@ export function PreviewConcierge(props: any) {
             }
             getHostToGuestBus().next({
               type: HOST_CHECK_IN,
-              payload: { editMode: false, highlightMode }
+              payload: { editMode: false, highlightMode, rteConfig: rteConfig.pageBuilder ?? {} }
```
Comment (jvega190) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]: @rart I added this for the pageBuilder rte to read a setup with id `pageBuilder`, is that ok? or should this one read base?

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 8]
```diff
@@ -374,7 +379,7 @@ export function PreviewConcierge(props: any) {
           if (type === GUEST_CHECK_IN) {
             getHostToGuestBus().next({
               type: HOST_CHECK_IN,
-              payload: { editMode: false, highlightMode }
+              payload: { editMode: false, highlightMode, rteConfig: rteConfig ?? {}, site: siteId }
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]: - We must resend rte config if it changes.
- Why are you sending the siteId now?

File: ui/guest/src/controls/rte.ts [hunk 9]
```diff
@@ -24,32 +24,72 @@ import { GuestStandardAction } from '../store/models/GuestStandardAction';
...
+import LookupTable from '@craftercms/studio-ui/models/LookupTable';
+import { reversePluckProps } from '../utils/object';
+// TODO: this needs to be imported from studio-ui
+import { SHOW_EDIT_DIALOG } from '../constants';
 
 export function initTinyMCE(
   record: ElementRecord,
-  validations: Partial<ContentTypeFieldValidations>
+  validations: Partial<ContentTypeFieldValidations>,
+  rteConfig?: LookupTable<any>,
```
Comment (rart) [inline ui/guest/src/controls/rte.ts]: You have enough information to create a proper type for rteConfig

File: ui/guest/src/controls/rte.ts [hunk 10]
```diff
@@ -24,32 +24,72 @@ import { GuestStandardAction } from '../store/models/GuestStandardAction';
...
+  const external = {
+    acecode: '/studio/static-assets/js/tinymce-plugins/ace/plugin.min.js',
+    editform: '/studio/static-assets/js/tinymce-plugins/editform/plugin.js'
+  };
+
+  if (rteConfig?.tinymceOptions?.external_plugins) {
+    Object.entries(rteConfig.tinymceOptions.external_plugins).forEach((entry) => {
+      external[entry[0]] = (entry[1] as string).replaceAll('{site}', site);
+    });
+  }
```
Comment (rart) [inline ui/guest/src/controls/rte.ts]: Expanding the `{site}` expression should be done at the reducer when parsing the config.
```suggestion
  const external = {
    ...rteConfig?.tinymceOptions?.external_plugins,
    acecode: '/studio/static-assets/js/tinymce-plugins/ace/plugin.min.js',
    editform: '/studio/static-assets/js/tinymce-plugins/editform/plugin.js'
  };
```

File: ui/guest/src/store/epics/root.ts [hunk 11]
```diff
@@ -340,7 +340,8 @@ const epic: Epic<GuestStandardAction, GuestStandardAction, GuestState> = combine
                     'Please add tinymce on to the page to enable editing.'
                 );
               } else if (not(validations?.readOnly?.value)) {
-                return initTinyMCE(record, validations);
+                const setupId = field.properties?.rteConfiguration?.value ?? 'generic';
+                return initTinyMCE(record, validations, state.rteConfig[setupId], state.activeSite);
```
Comment (rart) [inline ui/guest/src/store/epics/root.ts]: if setup is not defined, should use whatever is available

File: ui/guest/src/store/models/GuestStore.ts [hunk 12]
```diff
@@ -64,6 +64,8 @@ export interface GuestState {
     currentElement: number;
     registryEntryIds: number[];
   };
+  rteConfig: LookupTable<any>;
```
Comment (rart) [inline ui/guest/src/store/models/GuestStore.ts]: type

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/modules/Preview/PreviewConcierge.tsx [hunk 13]
```diff
@@ -277,6 +283,12 @@ export function PreviewConcierge(props: any) {
 
   // endregion
 
+  // region Update rte config
+  useEffect(() => {
+    getHostToGuestBus().next({ type: UPDATE_RTE_CONFIG, payload: { rteConfig } });
```
Comment (rart) [inline ui/app/src/modules/Preview/PreviewConcierge.tsx]:
```suggestion
  if (rteConfig)
    getHostToGuestBus().next({ type: UPDATE_RTE_CONFIG, payload: { rteConfig } });
```

**Review by rart — CHANGES_REQUESTED** [review]

**Review by rart — APPROVED** [review]

**Pull request merged**